### PR TITLE
cmd/protoc-gen-go-grpc: add protoc change detector test

### PR DESCRIPTION
**Pull request by zedGGs**
Heey ! 
Im adding this pull request for this issue: https://github.com/grpc/grpc-go/issues/6748

I will add more detailed explanation, thank you a lot for the time taken ! 

little gentle ping for @dfawley 
and one for @arvindbr8 

--------------------------------------------------------------------------------------------

NOTICE: there's a little issue with added `test_unary.proto` file maybe in `vet.sh`

I can update what's required

--------------------------------------------------------------------------------------------

The goal of this pr is to add test coverage for using 
`protoc-gen-go-grpc` internaly and compare results with usage of actual `protoc —go-grpc=.out test.proto`
from `google.golang.org/protobuf/compiler/protogen`

--------------------------------------------------------------------------------------------

A)

I have added tests for unary examples, but maybe it's good idea to make sure is this even good approach, 
including my approach in tests, function naming, const naming, too much init functions ? 
Is this code in tests reusable, for adding future examples of unary rpc proto file test examples

Can something be better in tests ? 

--------------------------------------------------------------------------------------------

Also I have added a little example of same test, but in other format. 
If you maybe prefer more this way.

--------------------------------------------------------------------------------------------

B)

One thing thats not the best, is `initPlugin` Compiler versions, 
Im using latest one, with `v4.24.3` but have no clue how to use or update to current ones in time, 
will take a closer look, so every time theres an update we don't have to update tests

--------------------------------------------------------------------------------------------

C) Added Small Update on this since `vet.sh` is greping these comments , Will fix error handling better 

Also I have not included “authors” comments that are generated above in pb.go files from proto files,
since `generateFile` function, does not do that, and if I want to do that
I can create helper function, that's gonna be called inside `generateFile` to do that job for me.
If this is even important.

--------------------------------------------------------------------------------------------

D)

By the way, we are only using here `protoc —go-grpc_out=. test.proto`, command that is used 
in wrapper function `generateFile`, that is using protogen, from `protobuf/compiler/protogen`

Is there maybe a plan of implementing `protoc —go_out=. test.proto` in other wrapper function,
for generating remaining pb.go file ? 
Does grpc.pb.go feels uncomplete without other pb.go file ? 

My plan is also to add additional tests, for all of helper functions that are being used.

--------------------------------------------------------------------------------------------
E)

Also I have already tested with sever streaming proto file, and seems to work, 
I can add that in this pull request as separated commit, 
or as totaly separated pull request ? 

Then we can add client and bidirectional streaming examples

--------------------------------------------------------------------------------------------
F)

Btw, do u prefer more to compare files that are being written to, and remove them with cleanup, 
or just to compare files content ? 

--------------------------------------------------------------------------------------------

G)

Other unary rpc test example: 
Do u prefer more naming of const with: dummySomething, mockSomething, or just smthng ? 

```
func TestGenerateFileUnaryRPC(t *testing.T) {
    testGeneratedGoProtoPath := "test_unary_grpc.pb.go"

    requireUnimplemented = proto.Bool(true)

    msg := []*descriptorpb.DescriptorProto{
        {
            Name: proto.String("EventRequest"),
            Field: []*descriptorpb.FieldDescriptorProto{
                {
                    Name:   proto.String("content"),
                    Type:   descriptorpb.FieldDescriptorProto_TYPE_BYTES.Enum(),
                    Number: proto.Int32(1),
                },
            },
        },
        {
            Name: proto.String("EventResponse"),
            Field: []*descriptorpb.FieldDescriptorProto{
                {
                    Name:   proto.String("content"),
                    Type:   descriptorpb.FieldDescriptorProto_TYPE_BYTES.Enum(),
                    Number: proto.Int32(1),
                },
            },
        },
    }

    mthd := []*descriptorpb.MethodDescriptorProto{
        {
            Name:       proto.String("addBlogPost"),
            InputType:  proto.String("EventRequest"),
            OutputType: proto.String("EventResponse"),
        },
    }

    svc := []*descriptorpb.ServiceDescriptorProto{
        {
            Name:   proto.String("BlogPostService"),
            Method: mthd,
        },
    }

    protoFile := []*descriptorpb.FileDescriptorProto{
        {
            Name:    proto.String("test_unary.proto"),
            Package: proto.String("main"),
            Options: &descriptorpb.FileOptions{
                GoPackage: proto.String("main/proto"),
            },
            MessageType: msg,
            Service:     svc,
        },
    }

    req := &pluginpb.CodeGeneratorRequest{
        ProtoFile:      protoFile,
        FileToGenerate: []string{"test_unary.proto"},
        CompilerVersion: &pluginpb.Version{
            Major: proto.Int32(4),
            Minor: proto.Int32(24),
            Patch: proto.Int32(3),
        },
    }

    plgn, err := protogen.Options{}.New(req)
    if err != nil {
        t.Errorf("Error while creating Plugin object: %v", plgn)
    }

    for _, f := range plgn.Files {
        if !f.Generate {
            continue
        }

        file := generateFile(plgn, f)
        fileContent, _ := file.Content()
        filePathUnary := strings.Replace(*f.Proto.Name, ".proto", "_grpc.pb.go", 1)

        err := os.WriteFile(filePathUnary, fileContent, 0644)
        if err != nil {
            t.Errorf("Error while writing go stub content to file: %s", err)
        }
    }

    testPluginContent, err := os.ReadFile(testGeneratedGoProtoPath)
    if err != nil {
        t.Errorf("Error while reading proto file")
    }

    protocPluginContent, err := os.ReadFile("testdata/proto/" + testGeneratedGoProtoPath)
    if err != nil {
        t.Errorf("Error while reading proto file")
    }

    if !reflect.DeepEqual(testPluginContent, protocPluginContent) {
        t.Errorf("Error generated go stubs are not equal")
    }

    t.Cleanup(func() {
        os.Remove(testGeneratedGoProtoPath)
    })
}
```

RELEASE NOTES: none

### CODE REVIEW

File: cmd/protoc-gen-go-grpc/grpc.go
Comment (zedGGs) [inline cmd/protoc-gen-go-grpc/grpc.go:537]: Will update this for sure

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (zedGGs) [inline cmd/protoc-gen-go-grpc/grpc_test.go:19]: This maybe can be an issue with up to date versions

**Comment by arvindbr8** [comment]
Thanks for sending us this PR @zedGGs. 

At a high level, we want to be testing the `protoc-gen-go-grpc` binary and not really unit test the behavior of `generateFile()`. There are a couple of ways of going about it, 
1. go test - call cmd.Command() from a go test to run `protoc-gen-go-grpc` with the input proto and check the output 
2. shell script - to build and exec the binary from the shell script. However inorder for the test to be run during CD we need to be able to call the shell script from a go test or create a new github workflow to call this shell script. 

I'm more inclined for (1). 

Let me know when you have the PR ready for review again and we can take another pass at it. 

Also please find my responses inline for the questions asked. 

Feel free to reach out if you have more questions.

----- 

> Heey ! Im adding this pull request for this issue: #6748
> 
> I will add more detailed explanation, thank you a lot for the time taken !
> 
> little gentle ping for @dfawley and one for @arvindbr8
> 
> NOTICE: there's a little issue with added `test_unary.proto` file maybe in `vet.sh`
> 
> I can update what's required

**arvindbr8:** `vet` is complaining because the `grpc_test.go` file that was added is missing the copyright message at the top. 

 
> The goal of this pr is to add test coverage for using `protoc-gen-go-grpc` internaly and compare results with usage of actual `protoc —go-grpc=.out test.proto` from `google.golang.org/protobuf/compiler/protogen`
> 
> A)
> 
> I have added tests for unary examples, but maybe it's good idea to make sure is this even good approach, including my approach in tests, function naming, const naming, too much init functions ? Is this code in tests reusable, for adding future examples of unary rpc proto file test examples

**arvindbr8:** I'm not sure if the code in this test might be reused in the future. Although inflating the init functions inplace might be better for readability, but I have no strong preference. Although this might not be required if we go with the new approach


> Can something be better in tests ?

**arvindbr8:** we should definitely add protos for all types of RPCs and tests as well. It might be easier to add this in this PR itself. 

> Also I have added a little example of same test, but in other format. If you maybe prefer more this way.
> 

**arvindbr8:** Could you link to that please? 

> B)
> 
> One thing thats not the best, is `initPlugin` Compiler versions, Im using latest one, with `v4.24.3` but have no clue how to use or update to current ones in time, will take a closer look, so every time theres an update we don't have to update tests
> 

**arvindbr8:** Not relevant if we go with the new approach


> C)
> 
> Also I have not included “authors” comments that are generated above in pb.go files from proto files, since `generateFile` function, does not do that, and if I want to do that I can create helper function, that's gonna be called inside `generateFile` to do that job for me. If this is even important.
> 

**arvindbr8:** Not relevant if we go with the new approach

> D)
> 
> By the way, we are only using here `protoc —go-grpc_out=. test.proto`, command that is used in wrapper function `generateFile`, that is using protogen, from `protobuf/compiler/protogen`
> 
> Is there maybe a plan of implementing `protoc —go_out=. test.proto` in other wrapper function, for generating remaining pb.go file ? Does grpc.pb.go feels uncomplete without other pb.go file ?
> 
> My plan is also to add additional tests, for all of helper functions that are being used.
> 

**arvindbr8:** `protoc` uses our `protoc-gen-go-grpc` when we the flag `--go-grpc-out` is being used. `protoc-gen-go-grpc` is the only binary we own in our repo. We do not have to test the behavior of other plugins here. 

We only want to be able to test changes made to `protoc-gen-go-grpc` binary.

> E)
> 
> Also I have already tested with sever streaming proto file, and seems to work, I can add that in this pull request as separated commit, or as totaly separated pull request ?
> 
> Then we can add client and bidirectional streaming examples


**arvindbr8:** can we add all of that in this same PR if thats not too much effort. Seems like we can reuse the same message and only add additional rpcs to the service. 

> F)
> 
> Btw, do u prefer more to compare files that are being written to, and remove them with cleanup, or just to compare files content ?
> 

**arvindbr8:** Let's remove them with cleanup

> G)
> 
> Other unary rpc test example: Do u prefer more naming of const with: dummySomething, mockSomething, or just smthng ?
> 
> ```
> func TestGenerateFileUnaryRPC(t *testing.T) {
>     testGeneratedGoProtoPath := "test_unary_grpc.pb.go"
> 
>     requireUnimplemented = proto.Bool(true)
> 
>     msg := []*descriptorpb.DescriptorProto{
>         {
>             Name: proto.String("EventRequest"),
>             Field: []*descriptorpb.FieldDescriptorProto{
>                 {
>                     Name:   proto.String("content"),
>                     Type:   descriptorpb.FieldDescriptorProto_TYPE_BYTES.Enum(),
>                     Number: proto.Int32(1),
>                 },
>             },
>         },
>         {
>             Name: proto.String("EventResponse"),
>             Field: []*descriptorpb.FieldDescriptorProto{
>                 {
>                     Name:   proto.String("content"),
>                     Type:   descriptorpb.FieldDescriptorProto_TYPE_BYTES.Enum(),
>                     Number: proto.Int32(1),
>                 },
>             },
>         },
>     }
> 
>     mthd := []*descriptorpb.MethodDescriptorProto{
>         {
>             Name:       proto.String("addBlogPost"),
>             InputType:  proto.String("EventRequest"),
>             OutputType: proto.String("EventResponse"),
>         },
>     }
> 
>     svc := []*descriptorpb.ServiceDescriptorProto{
>         {
>             Name:   proto.String("BlogPostService"),
>             Method: mthd,
>         },
>     }
> 
>     protoFile := []*descriptorpb.FileDescriptorProto{
>         {
>             Name:    proto.String("test_unary.proto"),
>             Package: proto.String("main"),
>             Options: &descriptorpb.FileOptions{
>                 GoPackage: proto.String("main/proto"),
>             },
>             MessageType: msg,
>             Service:     svc,
>         },
>     }
> 
>     req := &pluginpb.CodeGeneratorRequest{
>         ProtoFile:      protoFile,
>         FileToGenerate: []string{"test_unary.proto"},
>         CompilerVersion: &pluginpb.Version{
>             Major: proto.Int32(4),
>             Minor: proto.Int32(24),
>             Patch: proto.Int32(3),
>         },
>     }
> 
>     plgn, err := protogen.Options{}.New(req)
>     if err != nil {
>         t.Errorf("Error while creating Plugin object: %v", plgn)
>     }
> 
>     for _, f := range plgn.Files {
>         if !f.Generate {
>             continue
>         }
> 
>         file := generateFile(plgn, f)
>         fileContent, _ := file.Content()
>         filePathUnary := strings.Replace(*f.Proto.Name, ".proto", "_grpc.pb.go", 1)
> 
>         err := os.WriteFile(filePathUnary, fileContent, 0644)
>         if err != nil {
>             t.Errorf("Error while writing go stub content to file: %s", err)
>         }
>     }
> 
>     testPluginContent, err := os.ReadFile(testGeneratedGoProtoPath)
>     if err != nil {
>         t.Errorf("Error while reading proto file")
>     }
> 
>     protocPluginContent, err := os.ReadFile("testdata/proto/" + testGeneratedGoProtoPath)
>     if err != nil {
>         t.Errorf("Error while reading proto file")
>     }
> 
>     if !reflect.DeepEqual(testPluginContent, protocPluginContent) {
>         t.Errorf("Error generated go stubs are not equal")
>     }
> 
>     t.Cleanup(func() {
>         os.Remove(testGeneratedGoProtoPath)
>     })
> }
> ```
>

**arvindbr8:** Not application if we take the new approach

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Please take a look at my comment above and let me know when its ready for review again.

**Comment by zedGGs** [comment]
> Please take a look at my comment above and let me know when its ready for review again.

@arvindbr8 thank you a lot for the awesome review and all the explanation and advices in general and in all inline comments 

I will go with first option as u advise, add all protos for different types of rpcs, use cleanup, all in same pr

**Comment by arvindbr8** [comment]
@zedGGs -- I guess your changes were not pushed to your branch? Please take a look

**Comment by zedGGs** [comment]
Heeya @arvindbr8 ! 
I have updated pr, tests are going well locally, but is there possibility that "protoc" plugin is not installed on these machines in pipeline or something similar to that ? 

Of course I can update tests again, I appreciate all your time for reviews ! 
Also, I apologise for late update

**Comment by dfawley** [comment]
> I have updated pr, tests are going well locally, but is there possibility that "protoc" plugin is not installed on these machines in pipeline or something similar to that ?

You can call `/vet.sh -install` to make our existing vet script install it, or you can do something similar elsewhere.  I'd prefer to have only one place where we download/install protoc, or at least one place where we define the version to download.

**Comment by zedGGs** [comment]
@dfawley , thank you for the feedback and explanation, I will do that

File: vet.sh
Comment (arvindbr8) [inline vet.sh:41]: @dfawley 's comment was actually referring to the `protoc` install thats done from source right below. You could just run `./vet.sh -install` - to install the version of protoc used by Github Actions.

Comment (zedGGs) [reply vet.sh]: @arvindbr8 thank you for detailed explanation, I have updated pipeline and few more things, to make this be good

**Comment by zedGGs** [comment]
@arvindbr8 thank you for more detailed explanation, I have updated pipeline and few more little things, 
also here's little ping for @dfawley 

thank you guys !

**Comment by zedGGs** [comment]
@arvindbr8 thank you for more detailed explanation, I have updated pipeline and few more little things,
also here's little ping for @dfawley

thank you guys !

File: vet.sh
Comment (arvindbr8) [inline vet.sh:48]: It's not evident to me why we need to build `protoc-gen-go-grpc` as part of tests/vet. Am I missing something here?

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
I'm worried that maybe the intend of few PR comments are getting lost in translation. I dont want to be wasting too much of your time here on this issue. Do you mind if i take a stab at mine/dfawley's comments and commit to your branch? @zedGGs

File: vet.sh
Comment (zedGGs) [reply vet.sh:48]: Heey @arvindbr8, I had this issue without using that: 
https://github.com/grpc/grpc-go/actions/runs/7776819179/job/21204407334, 

its pipeline that was running without updated `vet.sh` to this change
```
 protoc-gen-go-grpc: program not found or is not executable
        Please specify a program using absolute path or make sure the program is available
```

even tho I have tried to add this in path using export, 
nothing seems to be good, and in `vet.sh` protoc is not being build to binary anywhere,
so only after this update I made this to be good

**Comment by zedGGs** [comment]
@arvindbr8 sure, feel free to add your commits, no worries u are not wasting my time at all

File: vet.sh
Comment (arvindbr8) [reply vet.sh:48]: Thanks for the clarification.

File: vet.sh
Comment (arvindbr8) [inline vet.sh:44]: Let's add a comment here to mention why we need to build `protoc-gen-go-grpc`

File: .github/workflows/testing.yml
Comment (arvindbr8) [inline .github/workflows/testing.yml:111]: Maybe we should just set `VET_SKIP_PROTO` for the "Run vet.sh" case and not set it at all for "Run tests"

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
LGTM, modulo 2 comments. PTAL

File: vet.sh
Comment (zedGGs) [reply vet.sh:44]: @arvindbr8 sure, make sense

File: .github/workflows/testing.yml
Comment (zedGGs) [reply .github/workflows/testing.yml:111]: @arvindbr8 sounds good, will update

File: regenerate.sh
Comment (arvindbr8) [inline regenerate.sh:31]: Let's update this comment to talk about the new pb.go's as well

**Review by arvindbr8 — APPROVED** [review]
LGTM modulo one comment. 

cc: @zasweq for second review

File: regenerate.sh
Comment (zedGGs) [reply regenerate.sh:31]: sure, that make sense

**Comment by arvindbr8** [comment]
@zedGGs - I would be cautious when I do a force push to your branch. Github UI handles that weirdly and sometimes have gotten rid of the previous conversations on the PR.

Also, Doug is on vacation and will be back next week.

**Comment by zedGGs** [comment]
@arvindbr8 sure, can agree on that, I think it's all ready and good at this point, 
maybe I can put some of our most important conversation here so its all at one place

**Review by arvindbr8 — APPROVED** [review]
I remember approving it already. Let's wait on dfawley's review before we can merge it

File: .github/workflows/testing.yml
Comment (dfawley) [inline .github/workflows/testing.yml:39]: This comment should be moved with the env lines and updated to something like `in the vet-proto job`.

File: .github/workflows/testing.yml
Comment (dfawley) [inline .github/workflows/testing.yml:112]: This doesn't want to be run for every version of Go that we test, presumably.

Let's test it in `extras` instead, below?

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc_test.go:30]: I don't think we need this `workingDir` at all.  `os.MkdirTemp` uses the default temp directory if the first parameter is `""` which seems perfectly fine and `command.Dir` defaults to the current directory already.

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc_test.go:68]: Please use `path.Join` instead.

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc_test.go:95]: Please use `cmp.Diff` and show the results so if things change we can see exactly what.

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (dfawley) [inline cmd/protoc-gen-go-grpc/grpc_test.go:83]: Same comment re: `path.Join` and below.

File: regenerate.sh
Comment (dfawley) [inline regenerate.sh:35]: Actually, I have renamed the files in `grpc_test_not_regenerate` to not be `.pb.go` anymore, so you can delete this part.

**Review by dfawley — COMMENTED** [review]
Thanks for the PR.. I have a few comments but overall this looks really good!

**Comment by arvindbr8** [comment]
Since this PR is going stale, I decided to open https://github.com/grpc/grpc-go/pull/7072 to attempt to fix it.

**Comment by zedGGs** [comment]
Heey @arvindbr8 , is there still possibility for me to finalise this pr ? I happen to be on vacation so I haven't reply on time, but Im available now

File: .github/workflows/testing.yml
Comment (zedGGs) [reply .github/workflows/testing.yml:39]: make sense, I can do that

File: .github/workflows/testing.yml
Comment (zedGGs) [reply .github/workflows/testing.yml:112]: sure, that's reasonable

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (zedGGs) [reply cmd/protoc-gen-go-grpc/grpc_test.go:30]: make sense, will do

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (zedGGs) [reply cmd/protoc-gen-go-grpc/grpc_test.go:68]: sounds like a plan

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (zedGGs) [reply cmd/protoc-gen-go-grpc/grpc_test.go:95]: sounds like an awesome plan

File: cmd/protoc-gen-go-grpc/grpc_test.go
Comment (zedGGs) [reply cmd/protoc-gen-go-grpc/grpc_test.go:83]: sure

File: regenerate.sh
Comment (zedGGs) [reply regenerate.sh:35]: will do

**Comment by arvindbr8** [comment]
Sorry about that @zedGGs. Please feel free to take up another issue in the repo. We would love to take a look at your PR.

**Pull request closed without merging**
